### ISSUE TITLE
Class definition in conditionals not correctly found

### ISSUE THREAD

**Issue by Flamefire** [pylint-dev/pylint#10238]
### Bug description

```python
I'm using a conditional type definition: A dataclass for Python 3.9+ and a namedtuple otherwise. Then I'm running PyLint with Python 3.12.

PyLint does not detect the unexpected and missing keyword args for that type.

Input:

# pylint: disable=missing-module-docstring
# pylint: disable=missing-class-docstring
# pylint: disable=too-few-public-methods
import sys
from dataclasses import dataclass

if sys.version_info >= (3, 9):
    @dataclass
    class Foo:
        x: int
        y: int
else:
    class Foo:
        pass


@dataclass
class Bar:
    x: int
    y: int


Foo(X2=1, y=2)
Bar(X2=1, y=2)


This is reduced, the actual fallback uses `Foo = namedtuple('Foo', ('x', 'y'))` but it shouldn't matter as that branch shouldn't be taken at all. Also swapping the branches and condition such that unused one comes first doesn't change the outcome
```

### Configuration

```ini

```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
/tmp/a.py:24:0: E1123: Unexpected keyword argument 'X2' in constructor call (unexpected-keyword-arg)
/tmp/a.py:24:0: E1120: No value for argument 'x' in constructor call (no-value-for-parameter)
```

### Expected behavior

The same error is shown for `Foo` and `Bar` on lines 23 & 24    
`Bar` in the example is only the sanity check for comparison, not relevant to the bug

### Pylint version

```shell
pylint 3.3.4
asteroid 3.3.8
Python 3.9.18
```

### Additional dependencies

```python

```
